### PR TITLE
bluetooth: fast_pair: Use per-connection authentication cbs

### PR DESCRIPTION
**Pull request by MarekPieta**
Change switches to using per-connection authentication callbacks.

Jira: NCSDK-16347

### CODE REVIEW

File: subsys/bluetooth/services/fast_pair/fp_auth.c [hunk 1]
```diff
@@ -245,6 +212,20 @@ int fp_auth_start(struct bt_conn *conn, bool send_pairing_req)
...
+	}
+
+	if (!auth_info_cb_registered) {
+		err = bt_conn_auth_info_cb_register(&conn_auth_info_callbacks);
+		if (err) {
+			return err;
+		}
+
+		auth_info_cb_registered = true;
+	}
```
Comment (kapi-no) [inline subsys/bluetooth/services/fast_pair/fp_auth.c:227]: Consider registering this using SYS_INIT. This way, you don't have to use static variable (and increase the number of state variables)

File: subsys/bluetooth/services/fast_pair/fp_auth.c [hunk 2]
```diff
@@ -245,6 +212,20 @@ int fp_auth_start(struct bt_conn *conn, bool send_pairing_req)
 	bt_auth_peer_passkeys[conn_idx] = EMPTY_PASSKEY;
 	update_conn_handling(conn, true);
 
+	err = bt_conn_auth_cb_overlay(conn, &conn_auth_callbacks);
+	if (err) {
+		return err;
+	}
```
Comment (kapi-no) [inline subsys/bluetooth/services/fast_pair/fp_auth.c:218]: Wouldn't it be better to overlay the callbacks first and then update the conn handling flag. This way, you will filter pairing attempts when the Fast Pair module isn't yet ready.

Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_auth.c:218]: That seems better approach. Changed. Thanks for the comment

**Review by KAGA164 — APPROVED** [review]

File: subsys/bluetooth/services/fast_pair/fp_auth.c (hunk 1 above)
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_auth.c:224]: sure

**Review by kapi-no — APPROVED** [review]

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 3]
```diff
@@ -109,22 +108,12 @@ No application input is required to handle the requests.
 Bluetooth authentication
 ========================
 
-The Fast Pair service overrides the set of Zephyr's Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`) while handling Key-Based Pairing request.
-The callbacks are updated to report proper device capabilities for Bluetooth LE pairing request/response packets.
-Overriding callbacks allows the GFPS to take over Bluetooth authentication during `Fast Pair Procedure`_ and perform all of the required operations without interacting with the application.
-After the procedure is finished, the previously used callbacks are assigned back.
+The Fast Pair service uses :c:func:`bt_conn_auth_cb_overlay` to overlay set of Zephyr's Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`) used for a connection if the connected Bluetooth Central uses Fast Pair pairing flow.
```
Comment (divipillai) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:111]:
```suggestion
The Zephyr's Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`) are used for a connection if the connected Bluetooth Central uses Fast Pair pairing flow.
Fast Pair service uses :c:func:`bt_conn_auth_cb_overlay` to overlay a set of Zephyr's Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`).
```

**Review by divipillai — COMMENTED** [review]
Shouldn't we have a change log entry?

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst (hunk 3 above)
Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:111]: That's only partially true. Let me rephrase the complex sentence in another way.

Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst]: I wonder if changelog entry is necessary here. Fast Pair is still in experimental state. This change does not affect library API too (the underlying implementation has changed, but the code still performs the same task).

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 4]
```diff
@@ -109,22 +108,14 @@ No application input is required to handle the requests.
 Bluetooth authentication
 ========================
 
-The Fast Pair service overrides the set of Zephyr's Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`) while handling Key-Based Pairing request.
-The callbacks are updated to report proper device capabilities for Bluetooth LE pairing request/response packets.
-Overriding callbacks allows the GFPS to take over Bluetooth authentication during `Fast Pair Procedure`_ and perform all of the required operations without interacting with the application.
-After the procedure is finished, the previously used callbacks are assigned back.
+The Bluetooth pairing is handled using a set of Bluetooth authentication callbacks (:c:struct:`bt_conn_auth_cb`).
+The pairing flow and the used set of Bluetooth authentication callbacks depends on connected peer:
```
Comment (greg-fer) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:112]:
```suggestion
The pairing flow and the set of Bluetooth authentication callbacks in use depend on the connected peer:
```

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 5]
```diff
@@ -109,22 +108,14 @@ No application input is required to handle the requests.
...
-In that case, the bonding uses Bluetooth authentication callbacks registered by the application.
-
-.. note::
-   Zephyr supports only one set of Bluetooth authentication callbacks.
-   Because of this limitation, it is not possible to simultaneously handle bonding using Fast Pair procedure and normal Bluetooth LE bonding.
-   In that case improper device capabilities would be reported.
-
-   Make sure that the normal Bluetooth LE bonding procedure and Fast Pair procedure do no overlap with each other.
-   The recommended approach is to avoid handling more than one bonding procedure at a time.
+* If the conneted peer follows Fast Pair pairing flow, Fast Pair service calls :c:func:`bt_conn_auth_cb_overlay` to automatically overlay the Bluetooth authentication callbacks.
```
Comment (greg-fer) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:114]:
```suggestion
* If the connected peer follows the Fast Pair pairing flow, the Fast Pair service calls :c:func:`bt_conn_auth_cb_overlay` to automatically overlay the Bluetooth authentication callbacks.
```

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 6]
```diff
@@ -109,22 +108,14 @@ No application input is required to handle the requests.
...
+  The function is called while handling Key-Based Pairing request.
```
Comment (greg-fer) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:115]:
```suggestion
  The function is called while handling the Key-Based Pairing request.
```

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 7]
```diff
@@ -109,22 +108,14 @@ No application input is required to handle the requests.
...
+  Overlying callbacks allows the GFPS to take over Bluetooth authentication during `Fast Pair Procedure`_ and perform all of the required operations without interacting with the application.
+* Otherwise normal Bluetooth LE pairing and global Bluetooth authentication callbacks are used.
```
Comment (greg-fer) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:117]:
```suggestion
* Otherwise, if the connected peer does not follow the Fast Pair pairing flow, normal Bluetooth LE pairing and global Bluetooth authentication callbacks are used.
```

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 8]
```diff
@@ -109,22 +108,14 @@ No application input is required to handle the requests.
...
+  This pairing flow is used, for example, if the connected peer does not support Fast Pair.
```
Comment (greg-fer) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:118]:
```suggestion
  This pairing flow is also used if the connected peer does not support Fast Pair.
```

Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:118]: I will rephrase this one a bit. `also` is a bit misleading here, since the peer which does not support Fast Pair does not follow Fast Pair pairing flow too.

Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:118]: Removed this sentence, I think it's not needed and it could cause some confusion

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 9]
```diff
@@ -109,22 +108,13 @@ No application input is required to handle the requests.
...
-
-.. note::
-   Zephyr supports only one set of Bluetooth authentication callbacks.
-   Because of this limitation, it is not possible to simultaneously handle bonding using Fast Pair procedure and normal Bluetooth LE bonding.
-   In that case improper device capabilities would be reported.
-
-   Make sure that the normal Bluetooth LE bonding procedure and Fast Pair procedure do no overlap with each other.
-   The recommended approach is to avoid handling more than one bonding procedure at a time.
+* If the connected peer follows the Fast Pair pairing flow, the Fast Pair service calls :c:func:`bt_conn_auth_cb_overlay` to automatically overlay the Bluetooth authentication callbacks.
+  The function is called while handling the Key-Based Pairing request.
```
Comment (divipillai) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:115]:
```suggestion
  The function is called while handling the Key-based Pairing request.
```
Reference: https://developers.google.com/nearby/fast-pair/specifications/service/gatt docs

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 10]
```diff
@@ -109,22 +108,13 @@ No application input is required to handle the requests.
...
+  Overlying callbacks allows the GFPS to take over Bluetooth authentication during `Fast Pair Procedure`_ and perform all of the required operations without interacting with the application.
+* Otherwise, if the connected peer does not follow the Fast Pair pairing flow, normal Bluetooth LE pairing and global Bluetooth authentication callbacks are used.
```
Comment (divipillai) [inline doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:117]: I think we could remove the bullets (*) and have a paragraph break after line 116.

**Review by divipillai — COMMENTED** [review]
Just some minor comments.

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst [hunk 11]
```diff
@@ -109,22 +108,13 @@ No application input is required to handle the requests.
...
-.. note::
-   Zephyr supports only one set of Bluetooth authentication callbacks.
-   Because of this limitation, it is not possible to simultaneously handle bonding using Fast Pair procedure and normal Bluetooth LE bonding.
-   In that case improper device capabilities would be reported.
-
-   Make sure that the normal Bluetooth LE bonding procedure and Fast Pair procedure do no overlap with each other.
-   The recommended approach is to avoid handling more than one bonding procedure at a time.
+* If the connected peer follows the Fast Pair pairing flow, the Fast Pair service calls :c:func:`bt_conn_auth_cb_overlay` to automatically overlay the Bluetooth authentication callbacks.
+  The function is called while handling the Key-Based Pairing request.
+  Overlying callbacks allows the GFPS to take over Bluetooth authentication during `Fast Pair Procedure`_ and perform all of the required operations without interacting with the application.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:116]: Are you sure that we should omit `s` in `allows`? `Overlying` is `it` (3-rd person), so shouldn't the `s` be here?

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst (hunk 10 above)
Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:117]: Changed

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:116]: Here we are talking about more than one Overlying callback (plural). Hence, it would be "allow" rather than "allows."

[the remark this answers is not in the document]
Comment (MarekPieta) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst]: Ok, I initially misunderstood you. Applied

File: doc/nrf/libraries/bluetooth_services/services/fast_pair.rst (hunk 10 above)
Comment (greg-fer) [reply doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:117]: I think a compromise could be having the following introductory sentence:
```
The pairing flow and the set of Bluetooth authentication callbacks in use depend on the connected peer and whether it follows the Fast Pair pairing flow:
```
And then the bullets introduced in similar fashion:
```
* If it follows the pairing flow, [...]
* If it does not follow the pairing flow, [...]
```
Personally, I always prefer bullets in similar situations, as they clearly show that some things are related to each other, while paragraphs might be about different topics. BUT two separate paragraphs would still be acceptable here. So it all depends on the phrasing and context.

**Review by greg-fer — APPROVED** [review]

**Review by divipillai — APPROVED** [review]
